src/QMCWaveFunctions/LCAO/MultiQuinticSpline1D.h

```diff
@@ -168,60 +168,88 @@ class MultiQuinticSpline1D
    * @param [out] u value of all splines at all electron distances [Nelec, Npbc, Nsplines]
    * @param Rmax spline will evaluate to zero for any distance greater than or equal to Rmax
   */
-  inline void batched_evaluate(const OffloadArray2D& r, OffloadArray3D& u, T Rmax) const
-  {
-    const size_t nElec = r.size(0);
-    const size_t Nxyz  = r.size(1); // number of PBC images
-    assert(nElec == u.size(0));
-    assert(Nxyz == u.size(1));
-    const size_t nRnl = u.size(2);    // number of splines
-    const size_t nR   = nElec * Nxyz; // total number of positions to evaluate
-
-    double one_over_log_delta = myGrid.OneOverLogDelta;
-    T lower_bound             = myGrid.lower_bound;
-    T log_delta               = myGrid.LogDelta;
-
-    auto* r_ptr = r.data();
-    auto* u_ptr = u.data();
-
-    auto* coeff_ptr       = coeffs_.data();
-    auto* first_deriv_ptr = first_deriv_.data();
-    const size_t nCols    = coeffs_.cols();
-    const size_t coefsize = coeffs_.size();
-    const int nsplines   = num_splines_;
+inline void batched_evaluate(const OffloadArray2D& r,
+                                   OffloadArray3D& u,
+                                   T Rmax) const
+{
+  // 1) Basic sizes
+  const size_t nElec = r.size(0);
+  const size_t Nxyz  = r.size(1);
+  assert(nElec == u.size(0));
+  assert(Nxyz == u.size(1));
+
+  const size_t nRnl  = u.size(2);       // number of splines
+  const size_t nR    = nElec * Nxyz;    // total [electron,image] positions
+  const size_t nsplines = num_splines_;
+  assert(nRnl == nsplines);
+
+  // 2) Obtain device pointers
+  const T* r_ptr        = r.device_data();
+  T*       u_ptr        = u.device_data();
+  const T* coeff_ptr    = coeffs_.device_data();
+  const T* fderiv_ptr   = first_deriv_.device_data();
+
+  // 3) Cache frequently used scalars
+  const double one_over_log_delta = myGrid.OneOverLogDelta;
+  const T      lower_bound        = myGrid.lower_bound;
+  const T      log_delta          = myGrid.LogDelta;
+  const size_t nCols              = coeffs_.cols();
```

Comment:
Using locals does help firstprivate scalars. It is not about cache.

Reply:
I am not understanding your comment...

Reply:
This is an optimization but the explanation is not correct.

Reply:
... We are at the level of correcting the comments!!??